File: problems/787/solution.go
package main
const MaxInt = int(^uint(0) >> 1)

func findCheapestPrice(n int, flights [][]int, src int, dst int, K int) int {
    ans, graph := MaxInt, map[int][][]int{}
    
    for _, flight := range flights {
        s := flight[0]
        graph[s] = append(graph[s], []int{flight[1], flight[2]}) 
    }
    
    dfs(graph, src, dst, 0, K + 1, &ans)
    
    if ans == MaxInt {
        return -1
    }
    
    return ans
}

func dfs(graph map[int][][]int, src, dst, p, k int, ans *int) {
    if k < 0 {
        return
    }
    
    if src == dst {
        *ans = min(*ans, p)
        return
    }
    
    for _, flight := range graph[src] {
        if flight[1] + p > *ans {
            continue
        }
        
        dfs(graph, flight[0], dst, p + flight[1], k - 1, ans)
    }
}

func min(a, b int) int {
    if a < b { return a }
    return b
}

File: problems/787/solution.py
# n = 3, edges = [[0,1,100],[1,2,100],[0,2,500]], src = 0, dst = 2, k = 1
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        ans, stops, adj, dists, q = float('inf'), 0, [[0 for _ in range(n)] for _ in range(n)], {}, deque([src])
        
        for f, t, p in flights:
            adj[f][t] = p
        
        dists[(src, 0)] = 0
        
        while q and stops <= k:
            qSize = len(q)
            
            for _ in range(qSize):
                node = q.popleft()
                
                for next in range(n):
                    if adj[node][next] > 0:
                        du = dists.get((node, stops), float('inf'))
                        dv = dists.get((next, stops + 1), float('inf'))
                        wuv = adj[node][next]
                        
                        if stops == k and next != dst:
                            continue
                        
                        if du + wuv < dv:
                            dists[next, stops + 1] = du + wuv
                            q.append(next)
                            
                            if next == dst:
                                ans = min(ans, du + wuv)
                                
            stops += 1
        
        return -1 if ans == float('inf') else ans